File: epi_judge_java/epi/ListCyclicRightShift.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class ListCyclicRightShift {

    @EpiTest(testDataFile = "list_cyclic_right_shift.tsv")
    public static ListNode<Integer> cyclicallyRightShiftList(ListNode<Integer> head, int k) {
        if (head == null) return null;

        int len = 0;
        var runner = head;
        var last = runner;

        while (runner != null) {
            if (runner.next == null) {
                last = runner;
            }

            runner = runner.next;
            len++;
        }

        k %= len;
        if (k == 0) return head;

        int shift = len - k;
        runner = head;
        for (int i = 1; i < shift; i++) {
            runner = runner.next;
        }

        var newHead = runner.next;
        runner.next = null;
        last.next = head;

        return newHead;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "ListCyclicRightShift.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/list_cyclic_right_shift.cc
#include <memory>

#include "list_node.h"
#include "test_framework/generic_test.h"

using std::shared_ptr;

shared_ptr<ListNode<int>> CyclicallyRightShiftList(shared_ptr<ListNode<int>> list, int k) {
    if (k == 0 || list == nullptr) return list;

    auto runner = list;
    int length = 0;

    shared_ptr<ListNode<int>> lastNode = nullptr;
    while (runner != nullptr) {
        length++;
        if (runner->next == nullptr) {
            lastNode = runner;
        }

        runner = runner->next;
    }

    k %= length;
    if (k == 0) return list;

    auto beforeNewHead = list;
    for (int i = 0; i < length - k - 1; i++) {
        beforeNewHead = beforeNewHead->next;
    }

    auto newHead = beforeNewHead->next;
    beforeNewHead->next = nullptr;
    lastNode->next = list;

    return newHead;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"L", "k"};
    return GenericTestMain(
            args, "list_cyclic_right_shift.cc", "list_cyclic_right_shift.tsv",
            &CyclicallyRightShiftList, DefaultComparator{}, param_names);
}
